common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java

```diff
@@ -1097,8 +1101,21 @@ public UTF8String copy() {
   @Override
   public int compareTo(@Nonnull final UTF8String other) {
     int len = Math.min(numBytes, other.numBytes);
-    // TODO: compare 8 bytes as unsigned long
-    for (int i = 0; i < len; i ++) {
+    int words = len / Longs.BYTES;
+    long roffset = other.getBaseOffset();
+    Object rbase = other.getBaseObject();
+    for (int i = 0; i < words * Longs.BYTES; i += Longs.BYTES) {
+      long left = getLong(base, offset + i);
+      long right = getLong(rbase, roffset + i);
+      if (left != right) {
+        if (!IS_LITTLE_ENDIAN) {
```

Comment:
Is it better to do this comparison once before executing a loop? (i.e. prepare two versions of this loop)

Reply:
@kiszk I think it'll result in exactly the same native code after JIT inlining the `static final boolean` check => I think this is cleaner since we don't need to duplicate the loop logic.

Reply:
`final` field can be updated by using reflection. Thus, I think that JIT compiler may be conservative for inlining regarding `static final`.

Reply:
@kiszk I don't think so, even if you update a primitive constant by reflection, all its inlined uses will not be updated. See http://java-performance.info/updating-final-and-static-final-fields/ who coded up an example of this at the bottom of the article.
Reflection updates will only update what you get back from reflection field access.

Reply:
(Nit: flip the condition to avoid the negation)
Is the point that if JIT thinks the value can't change then either way it can snip out the branch that will never happen? then yes checking the condition in or outside the loop doesn't matter. If that not always true, then I'm pretty clear that checking the condition inside this hot loop is slower.

My instinct would be to check it outside the loop just in case. However, I also note that it only matters once in the execution of the method, because as soon as `left != right`, the loop terminates. I think it's OK this way, therefore.

Reply:
Constants ("static final") are actually inlined by javac, not the JIT. Try it out.

```
class A {
  static final int CONSTANT = 1;
}

class B {

  B() { System.out.println(A.CONSTANT); }

}
```

```
$ javap -c B
Compiled from "foo.java"
class B {
  B();
    Code:
       0: aload_0
       1: invokespecial #1                  // Method java/lang/Object."<init>":()V
       4: getstatic     #2                  // Field java/lang/System.out:Ljava/io/PrintStream;
       7: iconst_1
       8: invokevirtual #4                  // Method java/io/PrintStream.println:(I)V
      11: return
}
```

Reply:
@vanzin it's a runtime constant but after the C2 pass it is completely removed by the compiler (just tried this out with Jitwatch).